### ISSUE TITLE
Support exporting and collecting metrics on GCP

### ISSUE THREAD

**Issue by tiziano88** [project-oak/oak#736]
At the very least we should get RPC stats (requests, responses, size, latency). Should be implemented directly on the Rust runtime (#540).

Probably need to be based on OpenCensus:

- https://cloud.google.com/monitoring/custom-metrics/open-census
- https://crates.io/crates/stackdriver_logger
- https://crates.io/crates/opentelemetry-stackdriver
- https://crates.io/crates/opentelemetry

As per #672, some of this may need to be controlled by a compile-time switch (e.g. `#ifdef OAK_DEBUG`), particularly if not aggregated.

**Comment by tiziano88** [comment]
I had a look around and it seems various pieces are available, but it's unclear how to put them all together, so that they work properly on GCP. @rbehjati could you take a closer look and advise on next steps?

**Comment by tiziano88** [comment]
I just came across https://crates.io/crates/tracing , it seems to have a nice API, @rbehjati do you have an opinion on it? Would it help with tracing / monitoring at all?

**Comment by rbehjati** [comment]
> I just came across https://crates.io/crates/tracing , it seems to have a nice API, @rbehjati do you have an opinion on it? Would it help with tracing / monitoring at all?

Thanks for the pointer @tiziano88. I experimented with it a bit. Here is an example of a log from a method with instrumentation:
```
Apr 09 00:45:23.054 DEBUG remove_reference{self=ChannelMapping { channels: RwLock { data: {3: Channel { messages: RwLock { data: [] }, writers: 0, readers: 1, waiting_threads: Mutex { data: {} }, label: Label { secrecy_tags: [], integrity_tags: [] } }, 1: Channel { messages: RwLock { data: [] }, writers: 1, readers: 1, waiting_threads: Mutex { data: {ThreadId(3): (Weak)} }, label: Label { secrecy_tags: [], integrity_tags: [] } }, 0: Channel { messages: RwLock { data: [] }, writers: 1, readers: 0, waiting_threads: Mutex { data: {ThreadId(3): (Weak)} }, label: Label { secrecy_tags: [], integrity_tags: [] } }, 2: Channel { messages: RwLock { data: [] }, writers: 1, readers: 1, waiting_threads: Mutex { data: {} }, label: Label { secrecy_tags: [], integrity_tags: [] } }} }, next_channel_id: 4, readers: RwLock { data: {Handle(14544449253528212536): 3, Handle(3332171126777907992): 2, Handle(5121763296857660169): 1} }, writers: RwLock { data: {Handle(14996249249004081182): 0, Handle(2042569363642778154): 1, Handle(2516576328973768130): 2} } } reference=Handle(14544449253528212536)}: oak_runtime::runtime::channel: remove_reference: deallocating channel 3
```
We could certainly use some of this additional information to filter the logs in Cloud monitoring, but I wonder if it is possible to get a better visualization too. 

The API is quite nice, especially the `#[tracing::instrument]` attribute. However to be able to use it, we need `formatting` for all the types (traits and structs) in the signature of the method we are instrumenting. This might be a limitation, unless we want to implement `std::fmt::Debug` for all the types.

Another potential drawback is that we have to replace the use of `log` with `tracing`, which means that the client applications won't be able to use implementations of the standard logger. But I guess this is not an issue, since only the oak-loader instantiates the runtime.

**Comment by tiziano88** [comment]
Neat! Can this integrate with https://cloud.google.com/trace/docs/quickstart too?

**Comment by rbehjati** [comment]
> Neat! Can this integrate with https://cloud.google.com/trace/docs/quickstart too?

I am not sure. I think for that we have to use opentelemetry-stackdriver. I'll have a closer look.

**Comment by tiziano88** [comment]
@rbehjati do you have an idea of what we would need at a minimum to support exporting metrics, for instance a simple counter of incoming gRPC requests? should we be looking at something [prometheus](https://prometheus.io/)? if so, would that require an HTTP server pseudo-node, or something along those lines?

**Comment by rbehjati** [comment]
> @rbehjati do you have an idea of what we would need at a minimum to support exporting metrics, for instance a simple counter of incoming gRPC requests? should we be looking at something [prometheus](https://prometheus.io/)? if so, would that require an HTTP server pseudo-node, or something along those lines?

If by "incoming gRPC requests" you mean requests processed by the gRPC server node, then I think the simplest solution is to log every incoming request, and count them using a log-based metric in GC Logging. For more granular metrics or to allow better filtering we could use [LogEntry](https://docs.rs/google-logging2/1.0.13+20200407/google_logging2/struct.LogEntry.html) with [RequestLog](https://cloud.google.com/logging/docs/reference/v1beta3/rest/v1beta3/RequestLog) as the `protobuf_payload`. I am looking into this these days. 

We can do this either by writing the logs to a file/stdout/stderr or by using the GC Logging REST API (for which we can use the [google_logging2](https://docs.rs/google-logging2/1.0.13+20200407/google_logging2/) crate). The cloud logging agent works with both modes of writing the logs. 

I am leaning towards using Google Cloud Operations for logging and monitoring for now. But I can imagine that some users might want to have the metrics in prometheus.  
In my opinion, we need an abstraction in the Runtime, and the SDK for logging and monitoring and two or three pluggable implementations (e.g., for prometheus and GC operations). 

If I understand correctly, for both prometheus and GC Logging REST API we'll need to use client libraries on the Oak side. So, we'll need an HTTP **client** pseudo-node. The server (either the GC logging agent or the prometheus server) is a non-Oak application. Is this correct?

BTW, prometheus has a [client library in Rust](https://github.com/tikv/rust-prometheus).

**Comment by tiziano88** [comment]
> If by "incoming gRPC requests" you mean requests processed by the gRPC server node, then I think the simplest solution is to log every incoming request, and count them using a log-based metric in GC Logging. For more granular metrics or to allow better filtering we could use [LogEntry](https://docs.rs/google-logging2/1.0.13+20200407/google_logging2/struct.LogEntry.html) with [RequestLog](https://cloud.google.com/logging/docs/reference/v1beta3/rest/v1beta3/RequestLog) as the `protobuf_payload`. I am looking into this these days.

Maybe that was too simplistic, let's say we would like to monitor request size and latency distribution. Basically the kind of things for which people usually rely on Prometheus.

> We can do this either by writing the logs to a file/stdout/stderr or by using the GC Logging REST API (for which we can use the [google_logging2](https://docs.rs/google-logging2/1.0.13+20200407/google_logging2/) crate). The cloud logging agent works with both modes of writing the logs.

I suggest separate the logging from monitoring aspects; we already have something working for logging (not great, but we can get logs out), while for monitoring we don't have anything yet. I don't think trying to emulate monitoring with logging would buy us much, strategically.

> I am leaning towards using Google Cloud Operations for logging and monitoring for now. But I can imagine that some users might want to have the metrics in prometheus.
> In my opinion, we need an abstraction in the Runtime, and the SDK for logging and monitoring and two or three pluggable implementations (e.g., for prometheus and GC operations).

What kind of abstraction would you suggest, and what changes do you think it would need in the Runtime? A special ABI call to export metrics, or a pseudo-node with special knowledge about the format of the metrics?

> If I understand correctly, for both prometheus and GC Logging REST API we'll need to use client libraries on the Oak side. So, we'll need an HTTP **client** pseudo-node. The server (either the GC logging agent or the prometheus server) is a non-Oak application. Is this correct?

My understanding is that this is not the case. Prometheus uses a [pull model](https://prometheus.io/docs/introduction/faq/#why-do-you-pull-rather-than-push), so it requires bringing up an HTTP **server**, and then a Prometheus metric collector scrapes the HTTP page(s) of the monitored tasks.

> BTW, prometheus has a [client library in Rust](https://github.com/tikv/rust-prometheus).

Nice, thanks!

**Comment by rbehjati** [comment]
Thanks. I did not know prometheus uses a pull model.

**Comment by rbehjati** [comment]
As discussed with @tiziano88, we'll focus on adding prometheus for now:
1. Start with a PoC for adding health-checks for the Runtime (e.g., up/down status, number of nodes). Requires adding a server that publishes the metrics on a configurable port. 
2. Extend it to metrics for the gRPC pseudo-node. Could be more generic and allow collecting and publishing metrics from all types of nodes.
3. Add a monitoring pseudo-node (or more generally an HTTP server pseudo-node) to allow application-level monitoring. 
4. Add a monitoring API to the SDK.
5. Hopefully there is a way to publish prometheus metrics on GCP. If not, we'll have to add support for that.

**Comment by tiziano88** [comment]
@rbehjati what is the next step for task? now that we have some metrics exported by the runtime, can we make a simple dashboard to show the exported values, and graph them over time? I think it would make for a nice demo!

**Comment by rbehjati** [comment]
I have started looking into the possible alternatives, and I have created a new task to work on it (#976). The problem is that it is not possible to decide which alternative is the best to use, without an example application that actually generates some metrics. Currently the aggregator example is not using the Rust gRPC server node, and according to @ipetr0v, the version that uses the Rust gRPC server node cannot be merged until C++ is completely discarded from Oak. 
I wonder if this is specific to the aggregator example or a general restriction for using gRPC? We could perhaps try and make a simple example application just for the sake of generating some metrics. Does that sound feasible?
